### PR TITLE
Multi-Threaded CPU Throughput Test, main branch (2022.11.24.)

### PR DESCRIPTION
**Pull request by krasznaa**
Re-wrote the "options code" a little, making it be more "according to my taste". :stuck_out_tongue: With this, the test introduced in #280 now outputs things in the following format:

```
[bash][atspot01]:build-x86_64_ubuntu2004_llvm_intel > ./bin/traccc_throughput_seq_example --detector_file=tml_detector/trackml-detector.csv --digitization_config_file=tml_detector/default-geometric-config-generic.json --input_directory=tml_full/ttbar_mu200/ --processed_events=10

Single-threaded throughput test

>>> Throughput options <<<
Input data format   : csv
Input directory     : tml_full/ttbar_mu200/
Detector geometry   : tml_detector/trackml-detector.csv
Digitization config : tml_detector/default-geometric-config-generic.json
Loaded event(s)     : 10
Processed event(s)  : 10

Reconstructed track parameters: 150357
Time totals:
                  File reading  5776 ms
              Event processing  1891 ms
Throughput:
              Event processing  189.163 ms/event, 5.28645 events/s
[bash][atspot01]:build-x86_64_ubuntu2004_llvm_intel >
```

Introduced a separate test that executes the "full reconstruction" on multiple threads using TBB, along with the "options code" necessary for it. It outputs information like:

```
[bash][atspot01]:build-x86_64_ubuntu2004_llvm_intel > ./bin/traccc_throughput_tbb_seq_example --detector_file=tml_detector/trackml-detector.csv --digitization_config_file=tml_detector/default-geometric-config-generic.json --input_directory=tml_full/ttbar_mu200/ --processed_events=100 --threads=8

Multi-threaded throughput test

>>> Throughput options <<<
Input data format   : csv
Input directory     : tml_full/ttbar_mu200/
Detector geometry   : tml_detector/trackml-detector.csv
Digitization config : tml_detector/default-geometric-config-generic.json
Loaded event(s)     : 10
Processed event(s)  : 100
>>> Multi-threading options <<<
CPU threads: 8

Reconstructed track parameters: 1883221
Time totals:
                  File reading  5729 ms
              Event processing  3891 ms
Throughput:
              Event processing  38.9144 ms/event, 25.6975 events/s
[bash][atspot01]:build-x86_64_ubuntu2004_llvm_intel >
```

I considered just updating @guilhermeAlmeida1's executable instead of introducing a separate one, but I thought this may be easier to start with. Especially since I wanted to show how I would propose to simplify the test executables. (They are a bit over-complicated at the moment in my mind.)

Eventually we should add some "status indicator" output to these executables, so that they would show nicely in longer tests that they are still working correctly. Since in longer tests right now one just has to trust the code that it's still doing things correctly, and didn't just get into an endless loop or something similar. But again, that's something for a future PR to handle.

### CODE REVIEW

**Review by guilhermeAlmeida1 — COMMENTED** [review]
Looks good to me, just one small comment

File: examples/options/src/options/mt_options.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    desc.add_options()(
+        "threads",
+        boost::program_options::value<std::size_t>()->default_value(1),
+        "The number of CPU threads to use");
+}
+
+void mt_options::read(const boost::program_options::variables_map& vm) {
+
+    threads = vm["threads"].as<std::size_t>();
+    if (threads == 0) {
```
[the remark this answers is not in the document]
Comment (guilhermeAlmeida1) [reply examples/options/src/options/mt_options.cpp:28]: It seems that TBB already sort of does this by changing this to the maximum if you more than that is requested, so I guess this should be fine as is

**Review by guilhermeAlmeida1 — APPROVED** [review]

File: examples/options/src/options/mt_options.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (krasznaa) [reply examples/options/src/options/mt_options.cpp:28]: Does it? :confused: I very much want to make it possible to over-subscribe our CPUs with these tests. We may actually see better performance that way in some of the tests.

So no, we should not add a maximum check. I was thinking of possibly adding a check for some ridiculous value (like 1000), but even decided against doing that in the end.

File: examples/options/src/options/throughput_options.cpp [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            throw std::invalid_argument("Invalid input data format specified");
+        }
+    }
+
+    // Set the rest of the options.
+    input_directory = vm["input_directory"].as<std::string>();
+    detector_file = vm["detector_file"].as<std::string>();
+    digitization_config_file = vm["digitization_config_file"].as<std::string>();
+    loaded_events = vm["loaded_events"].as<std::size_t>();
+    processed_events = vm["processed_events"].as<std::size_t>();
```
Comment (krasznaa) [inline examples/options/src/options/throughput_options.cpp:59]: I was instead pondering about adding some checks here. :thinking: But in the end decided against that... (Misconfigurations in these values should be fairly obvious as soon as the job starts running.)

File: examples/options/src/options/mt_options.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (guilhermeAlmeida1) [reply examples/options/src/options/mt_options.cpp:28]: As an example of trying to over-subscribe the CPU I get this message:

```
>>> Multi-threading options <<<
CPU threads: 8

TBB Warning: The number of workers is currently limited to 7. The request for 8 workers is ignored. Further requests for more workers will be silently ignored until the limit changes.
```

**Pull request merged**